### DISCUSSION TITLE
ROTATION AND STRETCHING ALIGNMENTS DURING  3D EBSD RECONSTRUCTION

### DISCUSSION THREAD

**Discussion by Tejanathreddy** [BlueQuartzSoftware/DREAM3DNX-Issues#23] [General]
I have been trying to reconstruct 3D EBSD using serial sectioning technique. I am facing issues with the EBSD .ang slices with rotation and stretching misalignment. I would like to modify the align section filter present in the DREAM3DNX to include the mentioned alignments. How can I modify the .cpp file of the Align section filter.

**Comment by nyoungbq** [comment]
Hello,

Our backend with the algorithms/filters is hosted here on Github at [simplnx](https://github.com/BlueQuartzSoftware/simplnx). The only caveat is that our front end is closed source so you will have to build from source and target NXRunner (which is a CLI pipeline runner) or use the python bindings. 

The align sections algorithm code is [here](https://github.com/BlueQuartzSoftware/simplnx/tree/develop/src/Plugins/SimplnxCore/src/SimplnxCore/Filters/Algorithms), there are two align sections algorithms, `AlignSectionsFeatureCentroid.cpp` and `AlignSectionsList.cpp`, so you will need to determine which one corresponds to the filter you are using.

If you can I would recommend just utilizing our python bindings to work with your data quickly for your use case. You can make entire python filters, or just make the modifications you need to the data after the first section of the pipeline, then use that data for the remainder of the pipeline. The documentation for which can be found at www.dream3d.io specifically [here](https://www.dream3d.io/python_docs/)

That said, if you prefer c++ and do make changes on your fork, and you are willing to contribute to the new/modified algorithms, open a PR against the main repository, and we will be happy to work to get it into develop.

I will be happy to answer any questions you have here, just `@` me, so I get notified.

**Reply by Tejanathreddy** [reply]
@nyoungbq Thanks for your reply.  I am using DREAM3DNX Version 7.0.0-RC-13 IN LINUX. The filter I would like to modify is Align sections misorientation filter. I would like to modify this filter to align sections wrt rotation and/or stretching. I think its better to modify individually to make the modificatiouns easier. I have tried to generate skeleton code by using the " Create python  plugins or filters"  by selecting 'use existing plugin'. I have created skeleton codes named rotation and stretching in .py format. while relaunching the DREAM3DNX I am getting a warning which I have attatched as a screen shot.
![Screenshot from 2024-09-13 11-52-34](https://github.com/user-attachments/assets/5382ef76-f1ee-45bb-95f1-1b18bb726275)
 Kindly help with this issue.

**Reply by nyoungbq** [reply]
Hey @Tejanathreddy,

I am not as experienced with the python bindings as @imikejackson, so I have handed the issue off to him. He dropped a response at the bottom of this thread to begin the debugging process with you. 

If we end up sort out the issue, it will mean a lot if you would mark the thread as answered for others who may run into similar issues. Thanks for reaching out, and I hope we can help you resolve this issue.

**Reply by Yu-1-maker** [reply]
Hey@Tejanathreddy ,
May I ask if you have resolved this issue and if the slice filter has been modified properly? During the process of reconstructing 3D EBSD, I encountered the same problem, which occurred due to rotation and stretching misalignment.
Thank you very much!!!!

**Reply by Tejanathreddy** [reply]
Hey @Yu-1-maker 
Unfortunately, I was not able to resolve this issue. As far as I know, there is a commercial software QUBE that has the capability of correcting translational and rotational/stretching corrections.  If you come across any other alternate solution, please let me know.

**Reply by Yu-1-maker** [reply]
Hey@Tejanathreddy ，
Regrettably, I am still a student at present and do not have the capability to obtain the commercial software QUBE. Are there any other methods for reconstructing 3D EBSD? If so, I would be extremely grateful!!!!

**Reply by Tejanathreddy** [reply]
Hey @Yu-1-maker
As far as I know, Dream3D is the only open source software which deals with 3D EBSD data formats as far as I know. Avizo is an alternative, but it typically works with preprocessed image stacks or tomography datasets.

**Reply by Yu-1-maker** [reply]
Hey @Tejanathreddy ,
I would like to ask you a question.I encountered confusion.During the 3D EBSD reconstruction process, I encountered a situation where the calculation of feature IDs was incorrect, which led to an error in the calculation of the IPF diagram.Is the reason for the error due to my data issue or the pipeline configuration problem?

My sample has two phases, BCC and FCC. Is this related to this?

The following three pictures respectively show my pipeline settings, incorrect illustrations, and the reasons for the errors.
<img width="643" height="1006" alt="32e13682ca1083e6a5f01e02e992f25b" src="https://github.com/user-attachments/assets/6ef790f8-9221-4ff8-a5ae-ae476680bb82" />
<img width="1280" height="414" alt="108b5fcc0044864b4e4eb833bb2ce5f1" src="https://github.com/user-attachments/assets/0e3efebc-af4a-46eb-bbe0-5fd6f79325cb" />
<img width="1263" height="794" alt="488b972cc004698015c2871ad76904d7" src="https://github.com/user-attachments/assets/5af105ce-957d-4d18-9910-1b11bbc8f3cf" />

If you could answer, I would be extremely grateful.

Many thanks,my dear friend.

**Comment by imikejackson** [comment]
@Tejanathreddy I would also like to add that in theory, you could create your own C++ plugin and as long as you compile yours the same way we compile DREAM3D-NX, then in theory you could use DREAM3D-NX and also include your own plugin with your version of the filter. That being said, we do not currently have good documentation on creating a filter from scratch in C++. 

Please do ask questions here and we will do our best to help you and improve the process where we can.

**Comment by imikejackson** [comment]
@Tejanathreddy When you set the path to the python plugin within DREAM3D-NX, you will need to make sure the path is set to an internal directory. For example if I have a plugin called NXDataAnalysisToolkit that I generated I would set the path to this inside of DREAM3D-NX:

```
/Users/mjackson/Workspace1/simplnx/wrapping/python/plugins/NXDataAnalysisToolkit/src/NXDataAnalysisToolkit
```
Please give that a try and let us know if that helps resolve the error.
